### PR TITLE
[Core][Geometries] Adding geometry order information

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

This PR defines a method in order to retrieve the geometry order of a geometry. This is a request by @jcotela to simplify its current workflow.

**🆕 Changelog**

- [Add `KratosGeometryOrderType`](https://github.com/KratosMultiphysics/Kratos/commit/4fe6b4e6373b8a404c8a4f26ea66754f3f95036d)
- [Define base `KratosGeometryOrderType`](https://github.com/KratosMultiphysics/Kratos/commit/01c98eedaa110d54fcb85d1848532f724f5967d6)
- [Add `Kratos_Zero_Order` geometries](https://github.com/KratosMultiphysics/Kratos/commit/e3a4668eaafd14f0ac4b340396abdf24a49e3adc)
- [Define `Kratos_Linear_Order` geometries](https://github.com/KratosMultiphysics/Kratos/commit/fb9b450b55b49561e60a98d693ffa0521c779665)
- [Add `Kratos_Quadratic_Order` geometries](https://github.com/KratosMultiphysics/Kratos/commit/b0bf48f49581a2dc24592ff82af05542a52c119b)
- [Adding `Kratos_Cubic_Order` geometries](https://github.com/KratosMultiphysics/Kratos/commit/659a8e4d4c76f0b0660523e1fb4da3c75f522bcc)
- [Adding `Kratos_Quartic_Order` geometries](https://github.com/KratosMultiphysics/Kratos/commit/97ea13d6a8d281982c0a4f5b475996314e47baeb)
- [Add some missing doc](https://github.com/KratosMultiphysics/Kratos/commit/4086f72c5fde947c29054801ad33bd0d7183cb17)

### CODE REVIEW

**Comment by loumalouomega** [comment]
@jcotela do we proceed?

File: kratos/geometries/geometry_data.h [hunk 1]
```diff
@@ -153,6 +152,29 @@ class GeometryData
...
+     * @var KratosGeometryOrderType::Kratos_Zero_Order: Defines a zero-order geometry. This is 0D geometry, aka discrete points, and is the simplest form of geometry.
+     * @var KratosGeometryOrderType::Kratos_Linear_Order: Defines a linear geometry. This is the simplest form, with minimal computational cost but lower accuracy.
+     * @var KratosGeometryOrderType::Kratos_Quadratic_Order: Defines a quadratic geometry. Offers a balance between computational cost and accuracy.
+     * @var KratosGeometryOrderType::Kratos_Cubic_Order: Defines a cubic geometry. Higher computational cost than quadratic, but offers improved accuracy.
+     * @var KratosGeometryOrderType::Kratos_Quartic_Order: Defines a quartic geometry. Further increases computational cost and accuracy.
+     * @var KratosGeometryOrderType::Kratos_Quintic_Order: Defines a quintic geometry. Represents the highest polynomial order here, with the highest accuracy and computational cost.
+     * @var KratosGeometryOrderType::Kratos_Unknown_Order: Used when the geometry order is unknown or not specified. Allows for flexibility in geometry definition.
+    */
+    enum KratosGeometryOrderType {
```
Comment (matekelemen) [inline kratos/geometries/geometry_data.h:168]: this will be **very** difficult to use. I'd rather recommend an optional int

Comment (loumalouomega) [reply kratos/geometries/geometry_data.h:168]: An enum can always be converted to a int, and the order coincides with the enum value

Comment (matekelemen) [reply kratos/geometries/geometry_data.h:168]: but why go the roundabout way of defining it as an `enum` in the first place? Why is an `enum` better here than an `std::optional<int>`? As far as I see it, the `enum` arbitrarily restricts the orders we can represent.

Comment (loumalouomega) [reply kratos/geometries/geometry_data.h:168]: We already use enums in many things for geometries, would you thinks it would be easier to just simply use unsigned int and end of story? opinions @KratosMultiphysics/technical-committee

Comment (matekelemen) [reply kratos/geometries/geometry_data.h:168]: > would you thinks it would be easier to just simply use unsigned int and end of story

an `unisgned` wouldn't be able to represent undefined order, and I think the undefined state of an `std::optional` fits better than using `-1` with signed integers, or some other arbitrary stuff.

Comment (loumalouomega) [reply kratos/geometries/geometry_data.h:168]: Okay

Comment (loumalouomega) [reply kratos/geometries/geometry_data.h:168]: @jcotela do you agree?, opinion  @KratosMultiphysics/technical-committee ?

Comment (jcotela) [reply kratos/geometries/geometry_data.h:168]: Hi @loumalouomega sorry for the delay, I had forgotten about this PR. Unfortunately, this change was intended for a feature that we won't be implementing so strictly speaking I no longer need it. Having said that, I think it may be useful to have a way to programmatically inquire about the interpolation order of a specific geometry instance, if the core developers agree.

Regarding the enum vs std::optional, I understand that the order is a category here (the geometry is lineal or quadratic in the same way it is a linear triangle or a cubic hexhahedron) so an enum seems appropriate. I could see using an int so that values have a natural greater/less than comparison in addition to equality checks, but I am not sure about the need for optional. After all, it should be straightforward to figure out the interpolation order for each derived geometry. @matekelemen what use case do you see for an undefined order? Working with base geometery instances?

Comment (loumalouomega) [reply kratos/geometries/geometry_data.h:168]: I will not touch this PR until a consencus is reached

Comment (matekelemen) [reply kratos/geometries/geometry_data.h:168]: > what use case do you see for an undefined order?

Well, there's a `Kratos_Unknown_Order` in the current PR so I thought there's a use case for it. Also geometries that don't overload this function should return an undefined value instead of some made-up default.

I'm not familiar with IGA in kratos, but I could imagine an undefined order for a default-constructed IGA geometry as well.

Comment (loumalouomega) [reply kratos/geometries/geometry_data.h:168]: If everyone agreess I can change that

**Comment by loumalouomega** [comment]
What is the status of this?

**Comment by matekelemen** [comment]
> What is the status of this?

I'm not blocking, just recommended an optinal int instead of an enum. I think we're still waiting for a review from the technical committee.

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee we should retake this

**Comment by matekelemen** [comment]
My comment is non-blocking, but my opinion still stands. Waiting for a review by the technical committee.

**Comment by rubenzorrilla** [comment]
> @KratosMultiphysics/technical-committee we should retake this

We'll have a look during next meeting.

**Comment by RiccardoRossi** [comment]
@KratosMultiphysics/technical-committee delegates the final decision about this to @jcotela

**Comment by loumalouomega** [comment]
Hello @jcotela?

**Review by jcotela — APPROVED** [review]
@loumalouomega sorry for the delay in answering. I'm a bit split on this. 

On the one hand, I think that what you implemented here is good to go. I am not convinced by the arguments for an optional int, since I do not see a use case where the integration order is undefined at run-time. I can see how it may be unknown at compile time, for example for nurbs-type geometries, but a given geometry instance should be able to determine its order. If the order can not be undefined at run-time, I would not burden the user of the method with checking for a valid optional.

On the other hand, I originally intended to use this to distinguish between linear and quadratic elements but we ended up not using it. I now believe that this was not the right choice for my problem. In particular, I'm thinking about Taylor-Hood P2P1 type elements, where being "linear" or "quadratic" depends on the DOF. In that case, only the element (actually the GeometricalObject in kratos terms) knows the right answer, and even then it can change for a given DOF.

So in summary, I think the code is good, but I am not sure we are addressing a real problem with this. If this was my work, I would close the PR to avoid adding dead code. I will approve in the hope that this is useful for someone else.
